### ISSUE TITLE
Use batch without data loader

### ISSUE THREAD

**Issue by allanj** [rusty1s/pytorch_geometric#179]
## ❓ Questions & Help

Just wondering how to use batch without the data loader. For example, my neural architecture is LSTM + GCN, then the GCN input is not directly from data (that's why I don't prefer to use data loader) but from LSTM.
The output from LSTM has the size of `(batch_size, number of nodes, Fin)`

Thus, in the example code, that is the size of my `x`. How do I efficiently adjust the code to fit in the batch size?
```
import torch
from torch.nn import Sequential as Seq, Linear as Lin, ReLU
from torch_geometric.nn import MessagePassing

class EdgeConv(MessagePassing):
    def __init__(self, F_in, F_out):
        super(EdgeConv, self).__init__(aggr='max')  # "Max" aggregation.
        self.mlp = Seq(Lin(2 * F_in, F_out), ReLU(), Lin(F_out, F_out))

    def forward(self, x, edge_index):
        # x has shape [N, F_in]
        # edge_index has shape [2, E]
        return self.propagate(edge_index, x=x)  # shape [N, F_out]

    def message(self, x_i, x_j):
        # x_i has shape [E, F_in]
        # x_j has shape [E, F_in]
        edge_features = torch.cat([x_i, x_j - x_i], dim=1)  # shape [E, 2 * F_in]
        return self.mlp(edge_features)  # shape [E, F_out]

```

**Comment by rusty1s** [comment]
Hi,

Generally, you can simply reshape your input tensor to `[batch_size * num_nodes, Fin]`. How is your `edge_index` given?

**Comment by allanj** [comment]
I still have a few questions.
1. Yes, I think you are right and I can do that. 
2. I was trying to understanding the `edge_index` with the following example in the documentation: [data handling of graphs](https://rusty1s.github.io/pytorch_geometric/build/html/notes/introduction.html#data-handling-of-graphs). 

In this case
```python3
import torch
from torch_geometric.data import Data

edge_index = torch.tensor([[0, 1, 1, 2],
                           [1, 0, 2, 1]], dtype=torch.long)
x = torch.tensor([[-1], [0], [1]], dtype=torch.float)

data = Data(x=x, edge_index=edge_index)
>>> Data(x=[3, 1], edge_index=[2, 4])
```

What I understand is:

- we have three nodes and their representations are `[-1]` for `node 0`, `[0]` for `node 1` and `[1] for node 2`, respectively
- The edges have two directions as the `edge_index` has the shape of `[2, number of edges]`. Here we have two edges
- The first direction is from `node 0 -> node 1` and `node 1 to node 2`.
- The second direction is symetric. 

**So if I am right**, in my case, for example my `batch size = 2`, also 3 nodes and two edges.
Then it should be something like the following after reshaping (I guess):
 ```python3
edge_index = torch.tensor([[0, 1, 1, 2, 3, 4, 4, 5],
                           [1, 0, 2, 1, 4, 3, 5, 4]], dtype=torch.long)
x = torch.tensor([[-1], [0], [1], [3], [-1], [0.4]], dtype=torch.float)
```
(I randomly set the representation of each node.)

But my question is **How to set the edge index, if I have some nodes in `x` is just padding?**

**Comment by rusty1s** [comment]
Your understanding is correct. That is how you data should look like. Given edge indices in a Python list, you can adjust them as follows (that's exactly what the data loader is doing internally):
```python
cumsum_nodes = 0
for i in range(num_nodes):
     edge_indices[i] += cumsum_nodes
     cumsum_nodes += num_nodes
edge_index = torch.cat(edge_indices, dim=1)
```
```
